tree/readspeed/src/ReadSpeed.cxx

```diff
@@ -0,0 +1,398 @@
+   }
+
+   return mergedClusters;
+}
+
+Result ReadSpeed::EvalThroughputMT(const Data &d, unsigned nThreads)
+{
+   ROOT::TThreadExecutor pool(nThreads);
+   const auto actualThreads = ROOT::GetThreadPoolSize();
+   if (actualThreads != nThreads)
+      std::cerr << "Running with " << actualThreads << " threads even though " << nThreads << " were requested.\n";
+
+   TStopwatch clsw;
+   clsw.Start();
+   const unsigned int maxTasksPerFile =
+      std::ceil(float(ROOT::TTreeProcessorMT::GetTasksPerWorkerHint() * actualThreads) / float(d.fFileNames.size()));
+
+   const auto rangesPerFile = MergeClusters(GetClusters(d), maxTasksPerFile);
+   clsw.Stop();
+
+   const size_t nranges =
+      std::accumulate(rangesPerFile.begin(), rangesPerFile.end(), 0u, [](size_t s, auto &r) { return s + r.size(); });
+   std::cout << "Total number of tasks: " << nranges << '\n';
+
+   auto treeIdx = 0;
+   std::vector<std::vector<std::string>> fileBranchNames;
+
+   std::vector<ReadSpeedRegex> regexes;
+   if (d.fUseRegex)
+      std::transform(d.fBranchNames.begin(), d.fBranchNames.end(), std::back_inserter(regexes), [](std::string text) {
+         return ReadSpeedRegex{text, std::regex(text)};
+      });
+
+   for (const auto &fName : d.fFileNames) {
+      std::vector<std::string> branchNames;
+      if (d.fUseRegex)
+         branchNames = GetMatchingBranchNames(fName, d.fTreeNames[treeIdx], regexes);
+      else
+         branchNames = d.fBranchNames;
+
+      fileBranchNames.push_back(branchNames);
+
+      if (d.fTreeNames.size() > 1)
+         ++treeIdx;
+   }
+
+   // for each file, for each range, spawn a reading task
+   auto sumBytes = [](const std::vector<ByteData> &bytesData) -> ByteData {
+      const auto uncompressedBytes =
+         std::accumulate(bytesData.begin(), bytesData.end(), 0ull,
+                         [](ULong64_t sum, const ByteData &o) { return sum + o.fUncompressedBytesRead; });
+      const auto compressedBytes =
+         std::accumulate(bytesData.begin(), bytesData.end(), 0ull,
+                         [](ULong64_t sum, const ByteData &o) { return sum + o.fCompressedBytesRead; });
+
+      return {uncompressedBytes, compressedBytes};
+   };
+
+   typedef std::map<int, TFile *> FileMap;
+   typedef std::map<std::thread::id, FileMap *> ThreadFileMap;
+   ThreadFileMap threadFileMap;
+
+   auto processFile = [&](int fileIdx) {
+      const auto &fileName = d.fFileNames[fileIdx];
+      const auto &treeName = d.fTreeNames.size() > 1 ? d.fTreeNames[fileIdx] : d.fTreeNames[0];
+      const auto &branchNames = fileBranchNames[fileIdx];
+
+      auto readRange = [&](const EntryRange &range) -> ByteData {
+         std::thread::id threadId = std::this_thread::get_id();
+         ThreadFileMap::iterator tMapIt = threadFileMap.find(threadId);
+         FileMap *fileMap;
+
+         if (tMapIt == threadFileMap.end()) {
+            fileMap = new FileMap();
+            threadFileMap.insert(std::make_pair(threadId, fileMap));
```

Comment:
this is not thread-safe, let's take a look at this logic together in person

Reply:
Ah fun, I thought ::insert was threadsafe but I see now it's not :smiling_face_with_tear:

Reply:
in C++ nothing is thread-safe unless it very explicitly is, because of the [zero-overhead principle](https://en.cppreference.com/w/cpp/language/Zero-overhead_principle)